errors/types.go

```diff
@@ -0,0 +1,186 @@
+	code int
+	// err is a nested error which is used to form a chain of errors for better context.
+	err error
+}
+
+// Kind determines the Kind associated with an error by inspecting the chain of errors. The top-most matching Kind is
+// returned or KindUnknown if no Kind can be determined.
+func Kind(err error) ErrKind {
+	var e CommonEdgex
+	if !errors.As(err, &e) {
+		return KindUnknown
+	}
+	// We want to return the first "Kind" we see that isn't Unknown, because
+	// the higher in the stack the Kind was specified the more context we had.
+	if e.kind != KindUnknown || e.err == nil {
+		return e.kind
+	}
+
+	return Kind(e.err)
+}
+
+// Error creates an error message taking all nested and wrapped errors into account.
+func (ce CommonEdgex) Error() string {
+	if ce.err == nil {
+		return ce.message
+	}
+
+	// ce.err.Error functionality gets the error message of the nested error and which will handle both CommonEdgex
+	// types and Go standard errors(both wrapped and non-wrapped).
+	return ce.message + " -> " + ce.err.Error()
+}
+
+// DebugMessages returns a string taking all nested and wrapped operations and errors into account.
+func (ce CommonEdgex) DebugMessages() string {
+	if ce.err == nil {
+		return ce.callerInfo + ": " + ce.message
+	}
+
+	if w, ok := ce.err.(CommonEdgex); ok {
+		return ce.callerInfo + ": " + ce.message + " -> " + w.DebugMessages()
+	} else {
+		return ce.callerInfo + ": " + ce.message + " -> " + ce.err.Error()
+	}
+}
+
+// Message returns the first level error message without further details.
+func (ce CommonEdgex) Message() string {
+	return ce.message
+}
+
+// Code returns the status code of this error.
+func (ce CommonEdgex) Code() int {
+	return ce.code
+}
+
+// Unwrap retrieves the next nested error in the wrapped error chain.
+// This is used by the new wrapping and unwrapping features available in Go 1.13 and aids in traversing the error chain
+// of wrapped errors.
+func (ce CommonEdgex) Unwrap() error {
+	return ce.err
+}
+
+// Is determines if an error is of type CommonEdgex.
+// This is used by the new wrapping and unwrapping features available in Go 1.13 and aids the errors.Is function when
+// determining is an error or any error in the wrapped chain contains an error of a particular type.
+func (ce CommonEdgex) Is(err error) bool {
+	switch err.(type) {
+	case CommonEdgex:
+		return true
+	default:
+		return false
+
+	}
+}
+
+// NewCommonEdgex creates a new CommonEdgex with the information provided.
+func NewCommonEdgex(kind ErrKind, message string, wrappedError error) CommonEdgex {
+	return CommonEdgex{
+		kind:       kind,
+		callerInfo: addCallerInformation(),
+		message:    message,
+		code:       codeMapping(kind),
+		err:        wrappedError,
+	}
+}
+
+// NewWrapperEdgex creates a new CommonEdgex to wrap another error to record the function call stacks.
+func NewWrapperEdgex(wrappedError error) CommonEdgex {
```

Comment:
Yes, we really need this.  Users might use NewCommonEdgex in the infrastructure layer (database access), and in the application layer, we returned the same error before.  However, using this `NewWrapperEdgex` could record the caller stack for helping debug instead of returning the original error directly .
For example, https://github.com/edgexfoundry/edgex-go/blob/b02541a61db9d4fcbee63351f76a5a0e52e2a310/internal/core/data/v2/application/event.go#L37

Reply:
renamed to `NewCommonEdgexWrapper`